File: android/app/src/main/java/com/hermexapp/android/features/onboarding/OnboardingViewModel.kt
package com.hermexapp.android.features.onboarding

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hermexapp.android.auth.AuthGateway
import com.hermexapp.android.auth.AuthManager
import com.hermexapp.android.model.AuthStatusResponse
import com.hermexapp.android.network.ApiError
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Port of the iOS `OnboardingViewModel`: drives the welcome → guidance →
 * connect flow, the test-connection probe, and password validation. All logic
 * lives in the suspend cores ([testConnectionNow]/[connectNow]) so JVM tests
 * exercise them directly; the UI goes through the launching wrappers.
 */
class OnboardingViewModel(
    private val authGateway: AuthGateway,
    savedServerUrl: String? = null,
    initialErrorMessage: String? = null,
) : ViewModel() {

    enum class Step { WELCOME, GUIDANCE, CONNECT }

    data class UiState(
        val step: Step = Step.WELCOME,
        val serverUrlString: String = "",
        val password: String = "",
        val authStatus: AuthStatusResponse? = null,
        val connectionMessage: String? = null,
        val errorMessage: String? = null,
        val isWorking: Boolean = false,
    ) {
        /**
         * No auth → no password. Passkey-only (auth on, password auth explicitly
         * off) → hide the field; connect() surfaces the unsupported message.
         * Unknown (null) keeps the show-the-field default. Mirrors iOS.
         */
        val isPasswordRequired: Boolean
            get() = authStatus?.authEnabled != false && authStatus?.passwordAuthEnabled != false
    }

    private val _uiState = MutableStateFlow(
        UiState(serverUrlString = savedServerUrl.orEmpty(), errorMessage = initialErrorMessage),
    )
    val uiState: StateFlow<UiState> = _uiState.asStateFlow()

    fun advanceToGuidance() = _uiState.update { it.copy(step = Step.GUIDANCE) }
    fun advanceToConnect() = _uiState.update { it.copy(step = Step.CONNECT) }
    fun backToWelcome() = _uiState.update { it.copy(step = Step.WELCOME) }
    fun backToGuidance() = _uiState.update { it.copy(step = Step.GUIDANCE) }

    fun updateServerUrl(value: String) = _uiState.update { it.copy(serverUrlString = value) }
    fun updatePassword(value: String) = _uiState.update { it.copy(password = value) }

    fun testConnection() {
        viewModelScope.launch { testConnectionNow() }
    }

    fun connect() {
        viewModelScope.launch { connectNow() }
    }

    suspend fun testConnectionNow() {
        _uiState.update { it.copy(errorMessage = null, connectionMessage = null, isWorking = true) }
        try {
            val status = authGateway.testConnection(_uiState.value.serverUrlString)
            val message = when {
                status.authEnabled == true && status.passwordAuthEnabled == false -> null
                status.authEnabled == true -> "Connection ok. Password required."
                else -> "Connection ok. Password not required."
            }
            val error = if (status.authEnabled == true && status.passwordAuthEnabled == false) {
                AuthManager.PASSKEY_ONLY_MESSAGE
            } else {
                null
            }
            _uiState.update {
                it.copy(authStatus = status, connectionMessage = message, errorMessage = error)
            }
        } catch (e: ApiError) {
            _uiState.update { it.copy(errorMessage = e.userMessage) }
        } finally {
            _uiState.update { it.copy(isWorking = false) }
        }
    }

    suspend fun connectNow() {
        _uiState.update { it.copy(errorMessage = null, connectionMessage = null) }

        passwordValidationMessage(_uiState.value.authStatus, _uiState.value.password)?.let { message ->
            _uiState.update { it.copy(errorMessage = message) }
            return
        }

        _uiState.update { it.copy(isWorking = true) }
        try {
            if (_uiState.value.authStatus == null) {
                val status = try {
                    authGateway.testConnection(_uiState.value.serverUrlString)
                } catch (e: ApiError) {
                    _uiState.update { it.copy(errorMessage = e.userMessage) }
                    return
                }
                _uiState.update { it.copy(authStatus = status) }

                passwordValidationMessage(status, _uiState.value.password)?.let { message ->
                    _uiState.update { it.copy(errorMessage = message) }
                    return
                }
            }

            authGateway.configure(_uiState.value.serverUrlString, _uiState.value.password)
            _uiState.update { it.copy(errorMessage = authGateway.lastErrorMessage.value) }
        } finally {
            _uiState.update { it.copy(isWorking = false) }
        }
    }

    companion object {
        /** Mirrors the iOS `passwordValidationMessage`. */
        fun passwordValidationMessage(authStatus: AuthStatusResponse?, password: String): String? {
            if (authStatus?.authEnabled != true) return null
            // Passkey-only servers don't take a password — configure() reports the
            // specific unsupported message instead of demanding one here.
            if (authStatus.passwordAuthEnabled == false) return null
            return if (password.trim().isEmpty()) AuthManager.EMPTY_PASSWORD_MESSAGE else null
        }
    }
}

File: HermesMobile/Features/Onboarding/OnboardingViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class OnboardingViewModel {
    nonisolated static let emptyPasswordMessage = String(localized: "Enter the server password.")

    var serverURLString = "" {
        didSet { invalidateProbedAuthStatusIfNeeded() }
    }
    var password = ""
    var customHeaders: [CustomHeader] = [] {
        didSet { invalidateProbedAuthStatusIfNeeded() }
    }
    var authStatus: AuthStatusResponse?
    var connectionMessage: String?
    var errorMessage: String?
    var isWorking = false

    @ObservationIgnored private var probedConnectionIdentity: String?
    @ObservationIgnored private var probeGeneration = 0

    init(
        savedServer: URL? = nil,
        savedHeaders: [CustomHeader] = [],
        initialErrorMessage: String? = nil
    ) {
        if let savedServer {
            serverURLString = savedServer.absoluteString
        }
        customHeaders = savedHeaders
        errorMessage = initialErrorMessage
    }

    var isPasswordRequired: Bool {
        // No auth → no password. Already signed in (trusted-header proxy) → no
        // password either. Passkey/OIDC-only → hide the field; connect()
        // surfaces the specific unsupported message instead. Unknown (nil)
        // keeps today's "show the field" default.
        guard authStatus?.authEnabled != false else { return false }
        guard authStatus?.isAlreadySignedIn != true else { return false }
        return authStatus?.passwordAuthEnabled != false
    }

    // MARK: - Connection identity (issue #285)

    private func currentConnectionIdentity() -> String {
        let urlPart: String
        do {
            let url = try AuthManager.normalizedServerURL(from: serverURLString)
            urlPart = url.absoluteString.lowercased()
        } catch {
            urlPart = serverURLString.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        }
        // Encode header names/values so delimiters (":" and "|") inside a
        // header value cannot collide with the identity serialization (e.g.
        // "X: 1|y:2" as one header vs "X: 1" + "Y: 2" as two). Values are not
        // lowercased — header values are case-sensitive.
        let headerPart = customHeaders.sanitizedForStorage()
            .map { header -> String in
                let name = Self.escapeForIdentity(header.sanitizedName.lowercased())
                let value = Self.escapeForIdentity(header.sanitizedValue)
                return "\(name):\(value)"
            }
            .sorted()
            .joined(separator: "|")
        return "\(urlPart)|\(headerPart)"
    }

    nonisolated private static func escapeForIdentity(_ raw: String) -> String {
        // Must escape "%" first so ":"/"|" escapes don't double-escape.
        var out = raw.replacingOccurrences(of: "%", with: "%25")
        out = out.replacingOccurrences(of: ":", with: "%3A")
        out = out.replacingOccurrences(of: "|", with: "%7C")
        return out
    }

    private func invalidateProbedAuthStatusIfNeeded() {
        guard let probed = probedConnectionIdentity else { return }
        let current = currentConnectionIdentity()
        if current != probed {
            authStatus = nil
            connectionMessage = nil
            errorMessage = nil
            probedConnectionIdentity = nil
        }
    }

    func testConnection(authManager: AuthManager) async {
        errorMessage = nil
        connectionMessage = nil
        isWorking = true
        defer { isWorking = false }

        probeGeneration += 1
        let generation = probeGeneration
        let identityAtStart = currentConnectionIdentity()

        do {
            let status = try await authManager.testConnection(
                serverURLString: serverURLString,
                customHeaders: customHeaders
            )
            guard generation == probeGeneration else { return }
            guard identityAtStart == currentConnectionIdentity() else { return }
            probedConnectionIdentity = identityAtStart
            authStatus = status
            if let message = AuthManager.unsupportedSignInMessage(for: status) {
                errorMessage = message
            } else if status.isAlreadySignedIn {
                connectionMessage = String(localized: "Connection ok. Already signed in by this server.")
            } else {
                connectionMessage = status.authEnabled == true
                    ? String(localized: "Connection ok. Password required.")
                    : String(localized: "Connection ok. Password not required.")
            }
        } catch {
            guard generation == probeGeneration else { return }
            guard identityAtStart == currentConnectionIdentity() else { return }
            probedConnectionIdentity = identityAtStart
            errorMessage = error.localizedDescription
        }
    }

    func connect(authManager: AuthManager) async {
        errorMessage = nil
        connectionMessage = nil

        if let validationMessage = Self.passwordValidationMessage(authStatus: authStatus, password: password) {
            errorMessage = validationMessage
            return
        }

        isWorking = true
        defer { isWorking = false }

        if authStatus == nil {
            probeGeneration += 1
            let generation = probeGeneration
            let identityAtStart = currentConnectionIdentity()
            do {
                let status = try await authManager.testConnection(
                    serverURLString: serverURLString,
                    customHeaders: customHeaders
                )
                guard generation == probeGeneration else { return }
                guard identityAtStart == currentConnectionIdentity() else { return }
                probedConnectionIdentity = identityAtStart
                authStatus = status
            } catch {
                guard generation == probeGeneration else { return }
                guard identityAtStart == currentConnectionIdentity() else { return }
                probedConnectionIdentity = identityAtStart
                errorMessage = error.localizedDescription
                return
            }

            if let validationMessage = Self.passwordValidationMessage(authStatus: authStatus, password: password) {
                errorMessage = validationMessage
                return
            }
        } else if probedConnectionIdentity == nil {
            // Auth status exists but probe identity was never recorded (e.g. restored
            // from a previous session). Seed it so future edits can invalidate.
            probedConnectionIdentity = currentConnectionIdentity()
        }

        await authManager.configure(
            serverURLString: serverURLString,
            password: password,
            customHeaders: customHeaders
        )
        errorMessage = authManager.lastErrorMessage
    }

    nonisolated static func passwordValidationMessage(authStatus: AuthStatusResponse?, password: String) -> String? {
        guard authStatus?.authEnabled == true else { return nil }
        // A server that already signed this client in (trusted-header proxy)
        // has no password to demand (#3).
        guard authStatus?.isAlreadySignedIn != true else { return nil }
        // Passkey/OIDC-only servers don't take a password either — let
        // configure() report the specific unsupported message instead of
        // demanding one here (#255, #3).
        guard authStatus?.passwordAuthEnabled != false else { return nil }

        let trimmedPassword = password.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmedPassword.isEmpty ? emptyPasswordMessage : nil
    }
}
